pkg/cloudevents/types/value.go

```diff
@@ -10,201 +10,212 @@ import (
-func (v Value) ToInteger() (int32, error) {
-	switch x := v.v.(type) {
+// ToInteger accepts any numeric value in int32 range, or canonical string.
+func ToInteger(v interface{}) (int32, error) {
+	v, err := Validate(v)
+	switch v := v.(type) {
 	case int32:
-		return x, nil
+		return v, nil
 	case string:
-		// Accept floating-point but truncate to int32 as per CE spec.
-		f, err := strconv.ParseFloat(x, 64)
-		if f > math.MaxInt32 || f < math.MinInt32 {
-			return 0, rangeErr(x)
-		}
-		return int32(f), err
+		return ParseInteger(v)
 	default:
-		return 0, typeErr("Integer", x)
+		return 0, convertErr("Integer", v, err)
 	}
 }
 
-// ToString returns a string value. It does NOT encode non-string, use String() for that.
-func (v Value) ToString() (string, error) {
-	switch x := v.v.(type) {
+// ToString returns a string value unaltered.
+//
+// This function does not perform canonical string encoding, use one of the
+// Format functions for that.
+func ToString(v interface{}) (string, error) {
+	v, err := Validate(v)
+	switch v := v.(type) {
 	case string:
-		return x, nil
+		return v, nil
 	default:
-		return "", typeErr("String", x)
+		return "", convertErr("String", v, err)
 	}
 }
 
-// ToBinary returns a []byte value, decoding from string if necessary.
-func (v Value) ToBinary() ([]byte, error) {
-	switch x := v.v.(type) {
+// ToBinary returns a []byte value, decoding from base64 string if necessary.
+func ToBinary(v interface{}) ([]byte, error) {
+	v, err := Validate(v)
+	switch v := v.(type) {
 	case []byte:
-		return x, nil
+		return v, nil
 	case string:
-		return base64.StdEncoding.DecodeString(x)
+		return base64.StdEncoding.DecodeString(v)
 	default:
-		return nil, typeErr("Binary", x)
+		return nil, convertErr("Binary", v, err)
 	}
 }
 
-// ToURIRef returns a url.URL value, decoding from string if necessary.
-func (v Value) ToURIRef() (url.URL, error) {
-	switch x := v.v.(type) {
-	case url.URL:
-		return x, nil
+// ToURL returns a *url.URL value, parsing from string if necessary.
+func ToURL(v interface{}) (*url.URL, error) {
+	v, err := Validate(v)
+	switch v := v.(type) {
+	case *url.URL:
+		return v, nil
 	case string:
-		u, err := url.Parse(x)
+		u, err := url.Parse(v)
 		if err != nil {
-			return url.URL{}, err
+			return nil, err
 		}
-		return *u, nil
+		return u, nil
 	default:
-		return url.URL{}, typeErr("URI-Reference", x)
+		return nil, convertErr("URL", v, err)
 	}
 }
 
-// ToTimestamp returns a Timestamp value, decoding from string if necessary.
-func (v Value) ToTimestamp() (time.Time, error) {
-	switch x := v.v.(type) {
+// ToTime returns a time.Time value, parsing from RFC3339 string if necessary.
+func ToTime(v interface{}) (time.Time, error) {
+	v, err := Validate(v)
+	switch v := v.(type) {
 	case time.Time:
-		return x, nil
+		return v, nil
 	case string:
-		ts, err := time.Parse(time.RFC3339Nano, x)
+		ts, err := time.Parse(time.RFC3339Nano, v)
```

Comment:
you could types.ParseTimestamp and then return the inner time.Time? Then the CE time format does not have to leak in two places.

Reply:
Wanted to bring that up - I don't like a ParseTimestamp that hides parse errors. I'd rather keep ParseTime and drop ParseTimestamp - the marshalling code can do
    ts, _ := ParseTime(s)
if it needs to ignore the error.

Reply:
ah interesting. As long as the time format parsing ends up in just one place, the type system seems like a good place.